File: java/diffie-hellman/src/main/java/DiffieHellman.java
import java.math.BigInteger;
import java.util.Random;

class DiffieHellman {
    private Random random = new Random();

    BigInteger privateKey(BigInteger prime) {
        BigInteger key;
        do {
            key = new BigInteger(prime.bitLength(), random);
        } while (key.compareTo(BigInteger.ONE) < 1 || key.compareTo(prime) >= 0);
        return key;
    }

    BigInteger publicKey(BigInteger p, BigInteger g, BigInteger privateKey) {
        return g.modPow(privateKey, p);
    }

    BigInteger secret(BigInteger p, BigInteger publicKey, BigInteger privateKey) {
        return publicKey.modPow(privateKey, p);
    }
}

File: kotlin/diffie-hellman/src/main/kotlin/DiffieHellman.kt
import java.math.BigInteger
import java.util.*

object DiffieHellman {
    fun privateKey(prime: BigInteger): BigInteger {
        val rand = Random()
        return generateSequence { BigInteger(prime.bitLength(), rand) }
                .filter { it > BigInteger.ONE && it < prime }
                .first()
    }

    fun publicKey(p: BigInteger, g: BigInteger, privKey: BigInteger): BigInteger
            = g.modPow(privKey, p)

    fun secret(prime: BigInteger, publicKey: BigInteger, privateKey: BigInteger): BigInteger
            = publicKey.modPow(privateKey, prime)
}
